### PR TITLE
Fix compilation warnings

### PR DESCRIPTION
**Pull request by asarium**
There are still a few warnings about deprecated functions but those are a bit harder to track down.

### CODE REVIEW

**Comment by chief1983** [comment]
Should probably get some convo going on this, I guess I could do some compile tests to make sure these don't break anything.  Other than that, anyone else have any complaints?

File: code/ship/ship.cpp [hunk 1]
```diff
@@ -10334,6 +10334,8 @@ int ship_fire_primary(object * obj, int stream_weapons, int force)
 				if (winfo_p->wi_flags3 & WIF3_APPLY_RECOIL){
 					firepoint_list = new vec3d[numtimes * points];
 					vm_vec_zero(&total_impulse);
+				} else {
+					firepoint_list = nullptr;
```
Comment (chief1983) [inline code/ship/ship.cpp]: This line is causing a compilation failure on Xcode - error: use of undeclared identifier 'nullptr'

Comment (niffiwan) [inline code/ship/ship.cpp]: What version of clang is Xcode using? nullptr is supposed to be supported from 3.0 according to [this page](http://clang.llvm.org/cxx_status.html)

Comment (chief1983) [inline code/ship/ship.cpp]: Kind of hard for me to tell for sure.  Here's a log excerpt:
```
CompileC build/FS2_Open.build/Debug/FS2_Open.build/Objects-normal/i386/plr_convert.o /Users/cliff.gordon/fs2_open_git/code/pilotfile/plr_convert.cpp normal i386 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch i386 -fmessage-length=218 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.6 -g -fvisibility=hidden -fvisibility-inlines-hidden -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/FS2_Open\ 3.7.3\ (debug).hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -I../../libjpeg -I../../lua -IFrameworks/SDL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/DerivedSources/i386 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/Objects-normal/i386/plr_convert.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/Objects-normal/i386/plr_convert.dia -c /Users/cliff.gordon/fs2_open_git/code/pilotfile/plr_convert.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/FS2_Open.build/Objects-normal/i386/plr_convert.o
```
I'm not sure what the '1.0' is at the beginning of that, but when I run the clang path with --version I get:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --version
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin12.5.0
Thread model: posix

Comment (niffiwan) [inline code/ship/ship.cpp]: To me that looks like it's based on clang 3.4. I guess the workaround is simple, revert to using the NULL macro, but it's a disappointing result.

Comment (asarium) [inline code/ship/ship.cpp]: Maybe C++11 support needs to be enabled with a compiler switch just like on linux?
[This SO answer](http://stackoverflow.com/a/7791906) says that it needs to be enabled in the project properties.

**Comment by chief1983** [comment]
Ok, attempting to set that requires changing the deployment target to 10.7, and would likely eliminate any remaining support for Xcode 3 and the old Xcode project files all in one swoop.  Not sure if we still compiled and ran on PPC but we definitely wouldn't after this.

**Comment by chief1983** [comment]
Also generates one of these per file:

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/ext/hash_map:212:5: warning: Use of the header <ext/hash_map> is deprecated. Migrate to <unordered_map> [-W#warnings]
#   warning Use of the header <ext/hash_map> is deprecated.  Migrate to <unordered_map>

**Comment by chief1983** [comment]
Here's the necessary changes to get it to compile.
```
diff --git a/projects/Xcode4/FS2_Open.xcodeproj/project.pbxproj b/projects/Xcode4/FS2_Open.xcodeproj/project.pbxproj
index 06a4269..f695249 100755
--- a/projects/Xcode4/FS2_Open.xcodeproj/project.pbxproj
+++ b/projects/Xcode4/FS2_Open.xcodeproj/project.pbxproj
@@ -4428,6 +4428,8 @@
                        isa = XCBuildConfiguration;
                        buildSettings = {
                                ARCHS = "$(ARCHS_STANDARD_32_64_BIT)";
+                               CLANG_CXX_LANGUAGE_STANDARD = "c++0x";
+                               CLANG_CXX_LIBRARY = "libc++";
                                CLANG_WARN_BOOL_CONVERSION = YES;
                                CLANG_WARN_CONSTANT_CONVERSION = YES;
                                CLANG_WARN_EMPTY_BODY = YES;
@@ -4458,7 +4460,7 @@
                                INFOPLIST_FILE = Info.plist;
                                INFOPLIST_PREPROCESS = YES;
                                LD_RUNPATH_SEARCH_PATHS = "@loader_path/../Frameworks";
-                               MACOSX_DEPLOYMENT_TARGET = 10.6;
+                               MACOSX_DEPLOYMENT_TARGET = 10.7;
                                PRODUCT_NAME = "FS2_Open $(CURRENT_PROJECT_VERSION) (debug)";
                                SDKROOT = macosx;
                                STRIP_INSTALLED_PRODUCT = NO;
@@ -4474,6 +4476,8 @@
                        isa = XCBuildConfiguration;
                        buildSettings = {
                                ARCHS = "$(ARCHS_STANDARD_32_64_BIT)";
+                               CLANG_CXX_LANGUAGE_STANDARD = "c++0x";
+                               CLANG_CXX_LIBRARY = "libc++";
                                CLANG_WARN_BOOL_CONVERSION = YES;
                                CLANG_WARN_CONSTANT_CONVERSION = YES;
                                CLANG_WARN_EMPTY_BODY = YES;
@@ -4507,7 +4511,7 @@
                                INFOPLIST_FILE = Info.plist;
                                INFOPLIST_PREPROCESS = YES;
                                LD_RUNPATH_SEARCH_PATHS = "@loader_path/../Frameworks";
-                               MACOSX_DEPLOYMENT_TARGET = 10.6;
+                               MACOSX_DEPLOYMENT_TARGET = 10.7;
                                PRODUCT_NAME = "FS2_Open $(CURRENT_PROJECT_VERSION)";
                                SDKROOT = macosx;
                                SEPARATE_STRIP = YES;
```

**Comment by asarium** [comment]
I applied your changes, it there anything else I need to change?

**Comment by chief1983** [comment]
Well, like I said, it generates a new warning for pretty much every single file now on OS X.  Is there a way to silence that or do what it says?  I feel like I remember some discussion about hash_map vs unordered_map and we went one way for a reason, but Xcode is complaining about it pretty heavily.

**Comment by asarium** [comment]
Looks like the clang check in vmallocator.h is faulty. We could use `__has_include` to check if `<unordered_map>` exists and fall back to <ext/hash_map> if it doesn't.

**Comment by Goober5000** [comment]
I'm in favor of not cutting out an entire swath of OS X machines.  But that's just me.

Is the Xcode issue a consequence of any C++11 support whatsoever, or just nullptr?

**Comment by chief1983** [comment]
Right now nullptr is the only one I've seen that requires this project file change.  But allowing any C++11 features means the change probably needs to be made, as one feature after another will probably need it.

**Comment by asarium** [comment]
I don't mind changing it to `NULL` but at some point C++11 needs to be enabled.

**Comment by Goober5000** [comment]
Butt that's what I'm referring to.  We already knew that enabling C++11 removes the ability to compile on VS 2005.  Now we've found out that it also removes the ability to compile on Xcode 3.

**Comment by asarium** [comment]
We already had this discussion. The decision to move C++11 was made and now we have to adjust the project files accordingly.

@chief1983: Given that we now have support for `std::unordered_map` we can use that everywhere instead of relying on the various hash map implementations. I'll add the required changes to this pull request.

**Comment by niffiwan** [comment]
With the latest commit, this now depends on #50 in order to compile on Linux.

**Comment by asarium** [comment]
It does? The branch compiled fine on my linux machine.

**Comment by niffiwan** [comment]
Interesting; I got this compile error with gcc 4.8.2;
```
In file included from /usr/include/c++/4.8/unordered_map:47:0,
                 from ./globalincs/vmallocator.h:12,
                 from ./globalincs/pstypes.h:645,
                 from ./missionui/redalert.h:15,
                 from missionui/redalert.cpp:12:
/usr/include/c++/4.8/bits/hashtable.h: In instantiation of ‘std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__node_type* std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_allocate_node(_Args&& ...) [with _Args = {const std::piecewise_construct_t&, std::tuple<const int&>, std::tuple<>}; _Key = int; _Value = std::pair<const int, int>; _Alloc = SCP_vm_allocator<int>; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<int>; _H1 = std::hash<int>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<false, false, true>; std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__node_type = std::__detail::_Hash_node<std::pair<const int, int>, false>]’:
/usr/include/c++/4.8/bits/hashtable_policy.h:493:8:   required from ‘std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::mapped_type& std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::operator[](const key_type&) [with _Key = int; _Pair = std::pair<const int, int>; _Alloc = SCP_vm_allocator<int>; _Equal = std::equal_to<int>; _H1 = std::hash<int>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<false, false, true>; std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::mapped_type = int; std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::key_type = int]’
/usr/include/c++/4.8/bits/unordered_map.h:596:20:   required from ‘std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::mapped_type& std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::operator[](const key_type&) [with _Key = int; _Tp = int; _Hash = std::hash<int>; _Pred = std::equal_to<int>; _Alloc = SCP_vm_allocator<int>; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::mapped_type = int; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::key_type = int]’
missionui/redalert.cpp:995:41:   required from here
/usr/include/c++/4.8/bits/hashtable.h:727:6: error: no matching function for call to ‘SCP_vm_allocator<std::__detail::_Hash_node<std::pair<const int, int>, false> >::construct(std::_Hashtable<int, std::pair<const int, int>, SCP_vm_allocator<int>, std::__detail::_Select1st, std::equal_to<int>, std::hash<int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::__node_type*&, const std::piecewise_construct_t&, std::tuple<const int&>, std::tuple<>)’
      _M_node_allocator().construct(__n, std::forward<_Args>(__args)...);
      ^
/usr/include/c++/4.8/bits/hashtable.h:727:6: note: candidate is:
In file included from ./globalincs/pstypes.h:645:0,
                 from ./missionui/redalert.h:15,
                 from missionui/redalert.cpp:12:
./globalincs/vmallocator.h:31:7: note: void SCP_vm_allocator<T>::construct(SCP_vm_allocator<T>::pointer, const T&) [with T = std::__detail::_Hash_node<std::pair<const int, int>, false>; SCP_vm_allocator<T>::pointer = std::__detail::_Hash_node<std::pair<const int, int>, false>*]
  void construct( pointer p, const T& value )
       ^
./globalincs/vmallocator.h:31:7: note:   candidate expects 2 arguments, 4 provided
make[1]: *** [missionui/redalert.o] Error 1

```

**Comment by asarium** [comment]
That's weird, I just tried a complete rebuild and it compiled fine (Ubuntu 15.04, GCC 4.9.2).

**Comment by niffiwan** [comment]
Very weird, I just retested with a fresh copy of this branch & a clean rebuild, and I got the same error.  I guess there must be some difference between 4.8.2 (ubuntu 14.04 lts) and 4.9.2. No big deal just need to wait for #50 to be merged before merging this one.

**Comment by chief1983** [comment]
This also isn't compiling on OS X for me right now.
```
CompileC build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.o /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=216 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -std=c++11 -stdlib=libc++ -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.7 -g -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/code.hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -IFrameworks/Ogg.framework/Headers -IFrameworks/Vorbis.framework/Headers -IFrameworks/Theora.framework/Headers -IFrameworks/SDL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenAL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources/x86_64 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.dia -c /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.o
In file included from /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:12:
In file included from ../../code/missionui/redalert.h:15:
In file included from ../../code/globalincs/pstypes.h:645:
In file included from ../../code/globalincs/vmallocator.h:12:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/unordered_map:724:5: error: static_assert failed "Invalid allocator::value_type"
    static_assert((is_same<value_type, typename allocator_type::value_type>::value),
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:12:
In file included from ../../code/missionui/redalert.h:15:
In file included from ../../code/globalincs/pstypes.h:645:
../../code/globalincs/vmallocator.h:116:34: note: in instantiation of template class 'std::__1::unordered_map<int, int, std::__1::hash<int>, std::__1::equal_to<int>, SCP_vm_allocator<int> >' requested here
class SCP_unordered_map : public std::unordered_map< Key, T, std::hash<Key>, std::equal_to<Key>, SCP_vm_allocator< T > > { };
                                 ^
/Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:825:30: note: in instantiation of template class 'SCP_unordered_map<int, int>' requested here
        SCP_unordered_map<int, int> Wing_pobjects_deleted;
                                    ^
/Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:826:31: error: no type named 'iterator' in 'SCP_unordered_map<int, int>'
        SCP_unordered_map<int, int>::iterator ii;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
2 errors generated.
```

**Comment by chief1983** [comment]
Also, that happens with PR #50 merged into this branch too, so not sure that's going to fix it.

**Comment by asarium** [comment]
I messed up the allocator declaration which may explain why the build broke for @niffiwan. I hope this also fixes this problem as `iterator` should always be a valid typedef.

**Comment by chief1983** [comment]
Ok, this compiles in Xcode now.

**Comment by asarium** [comment]
I fixed the compilation error introduced by the shadows merge and also fixed a new warning.
I think this is ready to be merged now.

**Comment by niffiwan** [comment]
darn, still fails to compile on gcc 4.8.2 unless #50 is merged first.  What about moving the hash_map -> unordered_map commit to it's own PR in order to get the rest of this committed without waiting for a solution to the VS2010 issue that's blocking #50?

**Comment by asarium** [comment]
I'm fine with that but that would mean OSX that users see a warning in every file that includes `vmallocator.h` (which is pretty much every file in the project). If @chief1983 is ok with that temporary situation until #50 is merged I'll open a new pull request.

**Comment by chief1983** [comment]
I can live with a warning, as long as it compiles.  Was #41 dependent on this one or that one?

**Comment by asarium** [comment]
Ok, I removed the unordered_map commit.
@niffiwan Can you check if it compiles without that?

**Comment by chief1983** [comment]
Compiled fine on 2010, FWIW.

**Comment by niffiwan** [comment]
yep, compiles fine on gcc 4.8.2. Together with @chief1983's confirmation that VS2010 is OK I think this can be merged

**Pull request merged**